src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -235,17 +236,35 @@ public boolean isFeedbackSessionHasQuestionForStudents(
     /**
      * Checks whether a student has completed a feedback session.
      *
-     * <p> If there is no question for students, the feedback session is completed</p>
+     * <p>If feedback session consists of all team questions, session is completed by student only
+     * if someone from the team has responded. If feedback session has some individual questions,
+     * session is completed only if the student has responded to the individual questions
+     * (regardless of the completion status of the team questions).</p>
      */
-    public boolean isFeedbackSessionCompletedByStudent(FeedbackSessionAttributes fsa, String userEmail) {
-        if (frLogic.hasGiverRespondedForSession(userEmail, fsa.getFeedbackSessionName(), fsa.getCourseId())) {
+    public boolean isFeedbackSessionCompletedByStudent(FeedbackSessionAttributes fsa, String userEmail, String userTeam) {
+        String feedbackSessionName = fsa.getFeedbackSessionName();
+        String courseId = fsa.getCourseId();
+
+        List<FeedbackQuestionAttributes> allQuestions =
+                fqLogic.getFeedbackQuestionsForStudents(feedbackSessionName, courseId);
+        // if there are no questions for student, session is complete
+        if (allQuestions.isEmpty()) {
             return true;
         }
+        boolean isAllTeamQuestions = true;
 
-        String feedbackSessionName = fsa.getFeedbackSessionName();
-        String courseId = fsa.getCourseId();
-        // if there is no question for students, session is complete
-        return !fqLogic.sessionHasQuestions(feedbackSessionName, courseId);
+        for (FeedbackQuestionAttributes question : allQuestions) {
+            if (question.getGiverType() == FeedbackParticipantType.STUDENTS) {
+                isAllTeamQuestions = false;
+                break;
+            }
+        }
+
+        if (isAllTeamQuestions) {
+            return frLogic.hasGiverRespondedForSession(userTeam, feedbackSessionName, courseId);
+        } else {
+            return frLogic.hasGiverRespondedForSession(userEmail, feedbackSessionName, courseId);
```

Comment:
If we trace `hasGiverRespondedForSession` until the DB layer, in `FeedbackResponseDb.java`, the `hasResponsesFromGiverInSession` method filters the responses by `giverEmail`. Passing the team name to `hasResponsesFromGiverInSession` does not seem to make sense as `giverEmail` is different from team name.

https://github.com/TEAMMATES/teammates/blob/d4dbaa26b27b0438b82467943d3726faf96e901e/src/main/java/teammates/storage/api/FeedbackResponsesDb.java#L175-L189

Reply:
Yea but from what I understand for team questions, the team name is what is actually stored in the db under `giverEmail`, which is why I initially suggest renaming the entity field. But I do agree there will prob be backward compatibility issues.

Reply:
Yes. The entity field is unfortunately `giverEmail`, possibly due to legacy issue. However, `FeedbackResponseAttribute` class has a more neutral field called `giver` with brief explanation about the nature of the field. We can also put some remark in the form of comment in the `FeedbackResponse` class to indicate this the same way the Attribute class does.

Reply:
The gitblame shows that the entity class is created back in 2013, and the renaming of `giverEmail` to `giver` in the Attribute class was performed in 2016 by PR #5547.

I am ok with renaming in code logic and adding comments in entity class to reflect this ambiguity.

Reply:
We can try fetching all students in the team and for each student check if the student has responded using `hasResponsesFromGiverInSession` return true if at least one of the students in the team has responded for the session.

Reply:
Yea that does seem a little expensive and will require storing the student email instead of the team name for team question responses. 
> We can also put some remark in the form of comment in the FeedbackResponse class to indicate this the same way the Attribute class does.

Personally I think this should be sufficient for now unless we need to track which team member answered the team questions.

Reply:
Slowly I feel like the cost might not justify the improvement of the UX as we moved away from key only query to reading many entities.